discovery/cvm/cvm.go

```diff
@@ -0,0 +1,305 @@
+	cvmLabelPublicIP      = cvmLabel + "public_ip"
+	cvmLabelVPCID         = cvmLabel + "vpc_id"
+	cvmLabelSubnetID      = cvmLabel + "subnet_id"
+	cvmLabelOsName        = cvmLabel + "os_name"
+	cvmLabelTag           = cvmLabel + "tag_"
+
+	Separator = ","
+
+	DefaultPageLimit = 100
+
+	EnvKeySecretId      = "PROMETHEUS_DS_CVM_SECRET_ID"
+	EnvKeySecretKey     = "PROMETHEUS_DS_CVM_SECRET_KEY"
+	EnvKeySecretToken   = "PROMETHEUS_DS_CVM_TOKEN"
+	EnvKeySecretRoleArn = "PROMETHEUS_DS_CVM_ROLE_ARN"
+)
+
+// DefaultSDConfig is the default CVM SD configuration.
+var DefaultSDConfig = SDConfig{
+	Port:            80,
+	RefreshInterval: model.Duration(60 * time.Second),
+}
+
+// Filter is the configuration for filtering CVM instances.
+type Filter struct {
+	Name   string   `yaml:"name"`
+	Values []string `yaml:"values"`
+}
+
+// SDConfig is the configuration for TencentCloud CVM based service discovery.
+type SDConfig struct {
+	Endpoint        string         `yaml:"endpoint,omitempty"`
+	Region          string         `yaml:"region,omitempty"`
+	Port            int            `yaml:"port"`
+	RefreshInterval model.Duration `yaml:"refresh_interval,omitempty"`
+	Filters         []*Filter      `yaml:"filters"`
+
+	// Tencent CVM Auth: https://github.com/tencentcloud/tencentcloud-sdk-go
+	SecretId  string `yaml:"secret_id,omitempty"`  // env set PROMETHEUS_DS_CVM_SECRET_ID
+	SecretKey string `yaml:"secret_key,omitempty"` // env set PROMETHEUS_DS_CVM_SECRET_KEY
+	Token     string `yaml:"token,omitempty"`      // env set PROMETHEUS_DS_CVM_TOKEN
+	RoleArn   string `yaml:"role_arn,omitempty"`   // env set PROMETHEUS_DS_CVM_ROLE_ARN
+}
+
+// Discovery periodically performs CVM-DescribeInstance requests. It implements
+// the Discoverer interface.
+type Discovery struct {
+	*refresh.Discovery
+	conf   *SDConfig
+	logger log.Logger
+}
+
+// UnmarshalYAML implements the yaml.Unmarshaler interface.
+func (c *SDConfig) UnmarshalYAML(unmarshal func(interface{}) error) error {
+	*c = DefaultSDConfig
+	type plain SDConfig
+	err := unmarshal((*plain)(c))
+	if err != nil {
+		return err
+	}
+
+	if c.Region == "" {
+		// todo get region from metadata api if region is blank
+		return errors.New("CVM SD configuration requires a region")
+	}
+	for _, f := range c.Filters {
+		if len(f.Values) == 0 {
+			return errors.New("CVM SD configuration filter values cannot be empty")
+		}
+	}
+	return nil
+}
+
+// NewDiscovery returns a new cvmDiscovery which periodically refreshes its targets.
+func NewDiscovery(conf *SDConfig, logger log.Logger) *Discovery {
+	if logger == nil {
+		logger = log.NewNopLogger()
+	}
+	d := &Discovery{
+		conf:   conf,
+		logger: logger,
+	}
+
+	d.Discovery = refresh.NewDiscovery(
+		logger,
+		"cvm",
+		time.Duration(conf.RefreshInterval),
+		d.refresh,
+	)
+	return d
+}
+
+func (d *Discovery) refresh(ctx context.Context) ([]*targetgroup.Group, error) {
+	_ = level.Debug(d.logger).Log("refresh with sd conf:", d.conf.Region, d.conf.Endpoint, d.conf.Filters)
```

Comment:
why the _ = ?

Reply:
just for suppressing warning "unhandled error"